packages/cli/src/sso/saml/routes/saml.controller.ee.ts

```diff
@@ -117,45 +115,61 @@ export class SamlController {
 		binding: SamlLoginBinding,
 	) {
 		try {
-			const loginResult = await this.samlService.handleSamlLogin(req, binding);
+			const { authenticatedUser, attributes, onboardingRequired } =
+				await this.samlService.handleSamlLogin(req, binding);
+
 			// if RelayState is set to the test connection Url, this is a test connection
 			if (isConnectionTestRequest(req)) {
-				if (loginResult.authenticatedUser) {
-					return res.send(getSamlConnectionTestSuccessView(loginResult.attributes));
+				if (authenticatedUser) {
+					return res.render('saml/connection-test-success.handlebars', {
+						attributes,
+						layout: 'saml',
+					});
 				} else {
-					return res.send(getSamlConnectionTestFailedView('', loginResult.attributes));
+					return res.render('saml/connection-test-failed.handlebars', {
+						attributes,
+						layout: 'saml',
+					});
 				}
 			}
-			if (loginResult.authenticatedUser) {
+
+			if (authenticatedUser) {
 				void Container.get(InternalHooks).onUserLoginSuccess({
-					user: loginResult.authenticatedUser,
+					user: authenticatedUser,
 					authenticationMethod: 'saml',
 				});
 				// Only sign in user if SAML is enabled, otherwise treat as test connection
 				if (isSamlLicensedAndEnabled()) {
-					await issueCookie(res, loginResult.authenticatedUser);
-					if (loginResult.onboardingRequired) {
+					await issueCookie(res, authenticatedUser);
+					if (onboardingRequired) {
 						return res.redirect(getInstanceBaseUrl() + SamlUrls.samlOnboarding);
 					} else {
 						return res.redirect(getInstanceBaseUrl() + SamlUrls.defaultRedirect);
 					}
 				} else {
-					return res.status(202).send(loginResult.attributes);
+					return res.status(202).send(attributes);
 				}
 			}
+
 			void Container.get(InternalHooks).onUserLoginFailed({
-				user: loginResult.attributes.email ?? 'unknown',
+				user: attributes.email ?? 'unknown',
 				authenticationMethod: 'saml',
 			});
+
 			throw new AuthError('SAML Authentication failed');
 		} catch (error) {
 			if (isConnectionTestRequest(req)) {
-				return res.send(getSamlConnectionTestFailedView((error as Error).message));
+				return res.render('saml/connection-test-failed.handlebars', {
```

Comment:
these will fail with `Cannot set headers after they are sent to the client`

could maybe be prevented by using callbacks in res.render